langchain4j-redis/src/main/java/dev/langchain4j/store/embedding/redis/RedisEmbeddingStoreImpl.java

```diff
@@ -0,0 +1,174 @@
+        String queryTemplate = "*=>[ KNN %d @%s $BLOB AS %s ]";
+        Query query = new Query(String.format(queryTemplate, maxResults, schema.getVectorFieldName(), RedisSchema.SCORE_FIELD_NAME))
+                .addParam("BLOB", ToByteArray(referenceEmbedding.vector()))
+                .returnFields(schema.getVectorFieldName(), schema.getScalarFieldName(), RedisSchema.SCORE_FIELD_NAME, schema.getMetadataFieldName())
+                .setSortBy(RedisSchema.SCORE_FIELD_NAME, true)
+                .dialect(2);
+
+        SearchResult result = client.ftSearch(schema.getIndexName(), query);
+        List<Document> documents = result.getDocuments();
+
+        return toEmbeddingMatch(documents);
+    }
+
+    private void createIndex(String indexName) {
+        if (!isIndexExist(indexName)) {
+            IndexDefinition indexDefinition = new IndexDefinition(IndexDefinition.Type.HASH);
+            indexDefinition.setPrefixes(schema.getPrefix());
+            String res = client.ftCreate(indexName, IndexOptions.defaultOptions()
+                    .setDefinition(indexDefinition), schema.toSchema());
+            if (!"OK".equals(res)) {
+                throw new JedisDataException("create index error, msg=" + res);
+            }
+        }
+    }
+
+    private boolean isIndexExist(String indexName) {
+        // redis do not contain command like ftExists
+        Set<String> indexSets = client.ftList();
+        return indexSets.contains(indexName);
+    }
+
+    private void addInternal(String id, Embedding embedding, TextSegment embedded) {
+        addAllInternal(Collections.singletonList(id), Collections.singletonList(embedding), embedded == null ? null : Collections.singletonList(embedded));
+    }
+
+    private void addAllInternal(List<String> ids, List<Embedding> embeddings, List<TextSegment> embedded) {
+        if (isCollectionEmpty(ids) || isCollectionEmpty(embeddings)) {
+            log.info("[do not add empty embeddings to elasticsearch]");
+            return;
+        }
+        ValidationUtils.ensureTrue(ids.size() == embeddings.size(), "ids size is not equal to embeddings size");
+        ValidationUtils.ensureTrue(embedded == null || embeddings.size() == embedded.size(), "embeddings size is not equal to embedded size");
+
+        int size = ids.size();
+        for (int i = 0; i < size; i++) {
+            String id = ids.get(i);
+            Embedding embedding = embeddings.get(i);
+            TextSegment textSegment = embedded == null ? null : embedded.get(i);
+            Map<byte[], byte[]> vectorField = new HashMap<>();
+            vectorField.put(schema.getVectorFieldName().getBytes(), ToByteArray(embedding.vector()));
+            if (textSegment != null) {
+                vectorField.put(schema.getScalarFieldName().getBytes(), textSegment.text().getBytes());
+                vectorField.put(schema.getMetadataFieldName().getBytes(), GSON.toJson(textSegment.metadata().asMap()).getBytes());
```

Comment:
we plan to add filtering by metadata soon. Is it possible to do filtering when all metadata is stored in a single field as a byte array?

Reply:
metadata filtering is not supported using byte arrays, now I use JSON to store data